### PR TITLE
UHF-2604: Ontology word details

### PR DESCRIPTION
**Pull request by tapsal**
Adds import for "ontology word details" and adds a new entity "OntologyWordDetails" for them. When these are imported, they are automatically published if the relevant Unit entity is also published.

For now, these are only used to get information high school search filtering.

**How to test:**
* Setup up Kasko development environment.
* Check out Kasko branch `UHF-2604-kasko-ontology_word_details`.
* Get the changes for this module: `composer require drupal/helfi_tpr:dev-UHF-2604-ontology_word_details`
* Run database updates: `drush updb`
* Import configs: `drush cim`
* Clear caches: `drush cr`
* Make sure you have imported tpr_units including Helsinki's upper secondary schools (you can use [this](https://www.hel.fi/helsinki/fi/kasvatus-ja-koulutus/lukiot/opiskelu/lukioiden-painotukset/) list to check what schools there are).
  * Publish those upper secondary school units (and maybe also e.g. Swedish translations)
* Make sure you have a landing page with High school search component (and maybe also a Swedish translation).
  * ...and those published schools are added to that search.
* Run: `drush migrate:import tpr_ontology_word_details`
* Now open the page containing the High school search and you should be able to filter the results using the "Select a special emphasis or study programme" and "Select a special educational mission" fields.
  * Using the list mentioned above, check that the filtering works correctly: correct emphasis and educational missions for correct schools.

Also check the relevant site specific PR: https://github.com/City-of-Helsinki/drupal-helfi-kasvatus-koulutus/pull/37

### CODE REVIEW

File: src/Plugin/views/filter/EducationalMission.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
+<?php
+
```
Comment (tuutti) [inline src/Plugin/views/filter/EducationalMission.php:2]: Missing `declare(strict_types=1)`

File: src/Plugin/views/filter/Emphasis.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
+<?php
```
Comment (tuutti) [inline src/Plugin/views/filter/Emphasis.php:1]: Missing declare(strict_types=1)

File: src/Plugin/views/filter/OntologyWordDetailsBase.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
+<?php
```
Comment (tuutti) [inline src/Plugin/views/filter/OntologyWordDetailsBase.php:1]: Missing declare(strict_types=1)

File: src/Plugin/Field/FieldFormatter/SchoolDetailsFormatter.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+   * {@inheritdoc}
+   */
+  public function viewElements(FieldItemListInterface $items, $langcode) : array {
+    /** @var \Drupal\helfi_tpr\Plugin\Field\FieldType\AccessibilitySentenceItem[] $items */
+    $elements = [];
+
+    foreach ($items as $delta => $item) {
+      $elements[$delta] = [
+        '#theme' => 'tpr-school-details',
+        '#clarification' => Html::escape($item->get('clarification')->getValue()),
```
Comment (tuutti) [inline src/Plugin/Field/FieldFormatter/SchoolDetailsFormatter.php:34]: This is escaped twice I think. Here and in twig template.

File: src/Plugin/Field/FieldFormatter/SchoolDetailsFormatter.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        '#schoolyear' => Html::escape($item->get('schoolyear')->getValue()),
```
Comment (tuutti) [inline src/Plugin/Field/FieldFormatter/SchoolDetailsFormatter.php:35]: Same as above.

**Review by tuutti — CHANGES_REQUESTED** [review]

File: src/Plugin/migrate/process/PublishByUnit.php [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+ * @endcode
+ */
+class PublishByUnit extends ProcessPluginBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function transform($value, MigrateExecutableInterface $migrate_executable, Row $row, $destination_property) {
+    $unit = Unit::load($value);
+    if (!empty($unit) && $unit->get('content_translation_status')->getString() === '1') {
```
Comment (tuutti) [inline src/Plugin/migrate/process/PublishByUnit.php:32]: Call `$unit->isPublished()` instead.

File: src/Entity/OntologyWordDetails.php [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  public static function baseFieldDefinitions(EntityTypeInterface $entity_type): array {
+    $fields = parent::baseFieldDefinitions($entity_type);
+
+    $fields['unit_id'] = static::createStringField('Unit ID')
+      ->setTranslatable(FALSE);
+
+    $fields['ontologyword_id'] = static::createStringField('Ontology word ID')
```
[the remark this answers is not in the document]
Comment (tapsal) [reply src/Entity/OntologyWordDetails.php:107]: No, the entity's ID field is just `id` (so it's generated by the RemoteEntityBase). These entities are basically a combination of units and ontologywords. However, as the migration source lacks a proper ID for that combination, the migration uses ontologyword_id and unit_id as source IDs and then the ID for the entity is generated based on that combination. This is done at the source plugin.

File: src/Entity/OntologyWordDetails.php [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+      ->setLabel(new TranslatableMarkup('School details'))
+      ->setRevisionable(FALSE)
+      ->setTranslatable(TRUE)
+      ->setCardinality(BaseFieldDefinition::CARDINALITY_UNLIMITED)
+      ->setDisplayConfigurable('view', TRUE)
+      ->setDisplayOptions('form', [
+        'type' => 'readonly_field_widget',
+      ]);
+
+    $fields['content_translation_source'] = BaseFieldDefinition::create('language')
```
[the remark this answers is not in the document]
Comment (tapsal) [reply src/Entity/OntologyWordDetails.php]: For some reason the content translation fields were not created when content translation support was enabled for this entity type. As the fields were previously created for other entity types using an update hook, these fields are now also now created by using another update hook.

File: src/Entity/OntologyWordDetails.php [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+      ->setLabel(t('Translation source'))
+      ->setDescription(t('The source language from which this translation was created.'))
+      ->setDefaultValue(LanguageInterface::LANGCODE_NOT_SPECIFIED)
+      ->setInitialValue(LanguageInterface::LANGCODE_NOT_SPECIFIED)
+      ->setRevisionable(TRUE)
+      ->setTranslatable(TRUE);
+
+    $fields['content_translation_outdated'] = BaseFieldDefinition::create('boolean')
```
[the remark this answers is not in the document]
Comment (tapsal) [reply src/Entity/OntologyWordDetails.php]: See comment above.

File: src/Entity/OntologyWordDetails.php [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+      ->setLabel(t('Translation outdated'))
+      ->setDescription(t('A boolean indicating whether this translation needs to be updated.'))
+      ->setDefaultValue(FALSE)
+      ->setInitialValue(FALSE)
+      ->setRevisionable(TRUE)
+      ->setTranslatable(TRUE);
+
+    $fields['content_translation_uid'] = BaseFieldDefinition::create('entity_reference')
```
[the remark this answers is not in the document]
Comment (tapsal) [reply src/Entity/OntologyWordDetails.php]: See comment above.

File: src/Entity/OntologyWordDetails.php [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+      ->setLabel(t('Translation author'))
+      ->setDescription(t('The author of this translation.'))
+      ->setSetting('target_type', 'user')
+      ->setSetting('handler', 'default')
+      ->setRevisionable(TRUE)
+      ->setDefaultValueCallback(ContentTranslationHandler::class . '::getDefaultOwnerId')
+      ->setTranslatable(TRUE);
+
+    $fields['content_translation_status'] = BaseFieldDefinition::create('boolean')
```
[the remark this answers is not in the document]
Comment (tapsal) [reply src/Entity/OntologyWordDetails.php]: See comment above.

File: src/Entity/OntologyWordDetails.php [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+      ->setLabel(t('Translation status'))
+      ->setDescription(t('A boolean indicating whether the translation is visible to non-translators.'))
+      ->setDefaultValue(TRUE)
+      ->setInitialValue(TRUE)
+      ->setRevisionable(TRUE)
+      ->setTranslatable(TRUE);
+
+    $fields['content_translation_created'] = BaseFieldDefinition::create('created')
```
[the remark this answers is not in the document]
Comment (tapsal) [reply src/Entity/OntologyWordDetails.php]: See comment above.

File: src/Entity/OntologyWordDetails.php [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+      ->setLabel(t('Translation created time'))
+      ->setDescription(t('The Unix timestamp when the translation was created.'))
+      ->setRevisionable(TRUE)
+      ->setTranslatable(TRUE);
+
+    $fields['content_translation_changed'] = BaseFieldDefinition::create('changed')
```
[the remark this answers is not in the document]
Comment (tapsal) [reply src/Entity/OntologyWordDetails.php]: See comment above.

**Review by tuutti — APPROVED** [review]

**Pull request merged**